src/python/alog/protocols.py

```diff
@@ -0,0 +1,576 @@
+        a true value, e.g.
+
+        logger.error("Houston, we have a %s", "major problem", exc_info=1)
+        """
+
+    # Aligned with the typeshed type definition:
+    # `*kwargs` replaced with: `exc_info`, `stack_info`, `stacklevel`, `extra`
+    def exception(
+        self,
+        msg: object,
+        *args: object,
+        exc_info: "_ExcInfoType" = True,
+        stack_info: bool = False,
+        stacklevel: int = 1,
+        extra: Optional[Mapping[str, object]] = None,
+    ) -> None:
+        """
+        Convenience method for logging an ERROR with exception information.
+        """
+
+    # Aligned with the typeshed type definition:
+    # `*kwargs` replaced with: `exc_info`, `stack_info`, `stacklevel`, `extra`
+    def critical(
+        self,
+        msg: object,
+        *args: object,
+        exc_info: "_ExcInfoType" = None,
+        stack_info: bool = False,
+        stacklevel: int = 1,
+        extra: Optional[Mapping[str, object]] = None,
+    ) -> None:
+        """
+        Log 'msg % args' with severity 'CRITICAL'.
+
+        To pass exception information, use the keyword argument exc_info with
+        a true value, e.g.
+
+        logger.critical("Houston, we have a %s", "major disaster", exc_info=1)
+        """
+
+    # Aligned with the typeshed type definition:
+    # `*kwargs` replaced with: `exc_info`, `stack_info`, `stacklevel`, `extra`
+    def log(
+        self,
+        level: int,
+        msg: object,
+        *args: object,
+        exc_info: "_ExcInfoType" = None,
+        stack_info: bool = False,
+        stacklevel: int = 1,
+        extra: Optional[Mapping[str, object]] = None,
+    ) -> None:
+        """
+        Log 'msg % args' with the integer severity 'level'.
+
+        To pass exception information, use the keyword argument exc_info with
+        a true value, e.g.
+
+        logger.log(level, "We have a %s", "mysterious problem", exc_info=1)
+        """
+
+    def _log(
+        self,
+        level: int,
+        msg: object,
+        args: "_ArgsType",
+        exc_info: Optional["_ExcInfoType"] = None,
+        extra: Optional[Mapping[str, object]] = None,
+        stack_info: bool = False,
+        stacklevel: int = 1,
+    ) -> None:
+        """
+        Low-level logging routine which creates a LogRecord and then calls
+        all the handlers of this logger to handle the record.
+        """
+
+    fatal = critical
+    if sys.version_info >= (3, 10):
+
+        def fatal(
+            self,
+            msg: object,
+            *args: object,
+            exc_info: "_ExcInfoType" = None,
+            stack_info: bool = False,
+            stacklevel: int = 1,
+            extra: Optional[Mapping[str, object]] = None,
+        ) -> None:
+            """
+            Don't use this method, use critical() instead.
+            """
+
+    def addHandler(self, hdlr: "Handler") -> None:  # pylint: disable=invalid-name
```

Comment:
It seems like these methods are all coming from `logging.Logger`. Is it possible to deduce a base for a `Protocol` from an existing class? ("Something that looks like this duck over here")

Reply:
Nope!

```
>>> from protocol_test import *
>>> class SomeProtocol(SomeClass, Protocol):
...     def baz(self, buz: int) -> None: ...
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/miniconda3/lib/python3.9/abc.py", line 106, in __new__
    cls = super().__new__(mcls, name, bases, namespace, **kwargs)
  File "/opt/miniconda3/lib/python3.9/typing.py", line 1251, in __init_subclass__
    raise TypeError('Protocols can only inherit from other'
TypeError: Protocols can only inherit from other protocols, got <class 'protocol_test.SomeClass'>
```